### ISSUE TITLE
VAE encoder silently corrupts latents on Apple Silicon (MPS): F.pad on dim 2 of a 5-D tensor returns zeros when H*W >= 65536

### ISSUE THREAD

**Issue by doggy8088** [QwenLM/Qwen-Image-2.1#6]
## Summary

On Apple Silicon (PyTorch MPS backend), `vae.encode()` silently returns corrupted latents for any input whose spatial size is about 512×512 or larger. The VAE **decoder** is unaffected, so text-to-image looks fine, but every **image-conditioned / editing** request is degraded: the edit itself comes out semantically correct — the requested change is applied and the composition and geometry are preserved — while colour and contrast collapse, giving a flat, desaturated, embossed-relief image. Nothing raises, and the latents stay finite, so there is no error to catch.

The cause is a PyTorch MPS bug in `F.pad` that `QwenImage21AvgDown3D` happens to trigger. A one-line change in the VAE works around it completely.

## Root cause

`QwenImage21AvgDown3D.forward` (`diffusers/models/autoencoders/autoencoder_kl_qwenimage21.py`) pads the temporal dimension of a 5-D tensor:

```python
pad_t = (self.factor_t - x.shape[2] % self.factor_t) % self.factor_t
pad = (0, 0, 0, 0, pad_t, 0)
x = F.pad(x, pad)          # <-- wrong result on MPS for large H*W
```

On the MPS backend, `F.pad` applied to dim 2 of a 5-D tensor returns a wrong result — usually **all zeros** — whenever the product of the last two dimensions is **≥ 65536 (2^16)**. It depends only on `H*W`, not on the channel count or the total element count, which suggests a 16-bit index or grid-size limit in the MPS kernel.

In the encoder, `down_blocks.1.avg_shortcut` has `factor_t=2, factor_s=2`, so `pad_t == 1` and the pad is taken. For a 512×512 image that shortcut receives `(1, 96, 1, 256, 256)` → `H*W = 65536` → the residual branch becomes exactly zero and everything downstream is garbage. A 448×448 image gives `224×224 = 50176 < 65536` and is fine, which is exactly the observed threshold.

Layer-by-layer tracing of the encoder against the CPU confirms the first divergence is `down_blocks.1.avg_shortcut`, whose MPS output is exactly all zeros while the CPU output has `std ≈ 1.27`. Its *input* matches the CPU to `2e-6`, so the input is fine and the op itself is at fault.

## Minimal reproducer 1 — the `F.pad` bug

```python
import torch, torch.nn.functional as F

torch.manual_seed(0)
for shape in [(1,  96, 1, 224, 224),   # H*W = 50176
              (1,   1, 1, 255, 256),   # H*W = 65280
              (1,   1, 1, 256, 256),   # H*W = 65536
              (1,  96, 1, 256, 256),   # H*W = 65536
              (1,   2, 1, 65536,  1),  # H*W = 65536
              (1, 512, 1, 128, 128)]:  # H*W = 16384, 8.4M elements
    x = torch.randn(*shape)
    ref = F.pad(x, (0, 0, 0, 0, 1, 0))                  # pad dim 2
    got = F.pad(x.to("mps"), (0, 0, 0, 0, 1, 0)).cpu()
    rel = ((ref - got).abs().max() / ref.abs().max()).item()
    print(f"{str(shape):>24} H*W={shape[3]*shape[4]:>7} rel_err={rel:.4f} "
          f"frac_zero={(got == 0).float().mean():.4f}")
```

```
    (1, 96, 1, 224, 224) H*W=  50176 rel_err=0.0000 frac_zero=0.5000   ok
     (1, 1, 1, 255, 256) H*W=  65280 rel_err=0.0000 frac_zero=0.5000   ok
     (1, 1, 1, 256, 256) H*W=  65536 rel_err=1.4691 frac_zero=0.5039   WRONG
    (1, 96, 1, 256, 256) H*W=  65536 rel_err=1.0000 frac_zero=1.0000   WRONG (all zeros)
     (1, 2, 1, 65536, 1) H*W=  65536 rel_err=1.0000 frac_zero=1.0000   WRONG (all zeros)
   (1, 512, 1, 128, 128) H*W=  16384 rel_err=0.0000 frac_zero=0.5000   ok
```

`frac_zero=0.5` is the correct value here, since padding `T: 1 → 2` makes half the output legitimately zero.

Further observations at `(1, 96, 1, 256, 256)`:

- Padding the **last** dimension (`(1,0,0,0,0,0)`) or the **H** dimension (`(0,0,1,0,0,0)`) is correct (`rel_err = 0`). Only the dim-2 pad breaks.
- Deterministic: the same input tensor gives `rel_err = 1.000000, frac_zero = 1.0000` on every run.
- `float32` gives all zeros; `float16` and `bfloat16` give **NaN**.
- `torch.cat([torch.zeros_like(x[:, :, :1]), x], dim=2)` is exact (`rel_err = 0.000000`).

## Minimal reproducer 2 — the resulting VAE corruption

Synthetic input, no image file needed:

```python
import os
os.environ.setdefault("PYTORCH_ENABLE_MPS_FALLBACK", "1")
import torch
from diffusers import AutoencoderKLQwenImage21

cpu = AutoencoderKLQwenImage21.from_pretrained(
    "Qwen/Qwen-Image-2.1", subfolder="vae", torch_dtype=torch.float32).eval()
mps = AutoencoderKLQwenImage21.from_pretrained(
    "Qwen/Qwen-Image-2.1", subfolder="vae", torch_dtype=torch.float32).eval().to("mps")

torch.manual_seed(0)
for size in (256, 448, 512, 768, 1024):
    x = torch.rand(1, 4, 1, size, size) * 2 - 1
    with torch.no_grad():
        a = cpu.encode(x).latent_dist.mode()
        b = mps.encode(x.to("mps")).latent_dist.mode().float().cpu()
    print(f"{size}x{size}: rel_err={(a - b).abs().max() / a.abs().max():.4f}")
```

```
256x256:   rel_err=0.0000
448x448:   rel_err=0.0000
512x512:   rel_err=0.2052
768x768:   rel_err=0.1519
1024x1024: rel_err=0.4582
```

This happens in `float32` as well as `bfloat16`, so it is not a precision or numerical-instability problem.

## Impact

Round-tripping a 768×1376 photograph through `vae.encode` → `vae.decode`, measured against the original:

| VAE device / dtype | RGB std | mean saturation | min alpha | result |
|---|---|---|---|---|
| original image | ~55 | 0.628 | 255 | — |
| CPU float32 | 57.4 | 0.630 | 253 | faithful |
| MPS float32 | 31.3 | 0.401 | 47 | corrupted |
| MPS bfloat16 | 9.1 | 0.155 | 141 | badly corrupted |

Encoding on CPU and decoding on MPS is faithful; encoding on MPS is corrupt regardless of where the decode runs. That isolates the fault to the encoder, and explains why text-to-image (decoder only) looks fine while every edit is broken.

Because this VAE is RGBA (`in_channels: 4`), the corruption also pushes the alpha channel away from opaque, so a pipeline that infers "is this transparent?" from the decoded alpha will wrongly ship a semi-transparent RGBA image. This is a downstream symptom of the same root cause rather than a second bug — see Verification below.

## Suggested fix

Replacing the temporal `F.pad` with `torch.cat` in `QwenImage21AvgDown3D.forward` fixes it completely, and is equivalent on CUDA and CPU:

```python
def forward(self, x: torch.Tensor) -> torch.Tensor:
    pad_t = (self.factor_t - x.shape[2] % self.factor_t) % self.factor_t
    if pad_t:
        # F.pad on dim 2 of a 5-D tensor is broken on the MPS backend when
        # H*W >= 65536; torch.cat is equivalent and correct everywhere.
        x = torch.cat(
            [x.new_zeros(x.shape[0], x.shape[1], pad_t, x.shape[3], x.shape[4]), x],
            dim=2,
        )
    B, C, T, H, W = x.shape
    ...
```

`QwenImage21DupUp3D` and any other 5-D `F.pad` call site are worth auditing for the same pattern.

## Verification

**1. VAE round trip**, 768×1376, MPS float32 — with the patch: `RGB std 57.4`, `saturation 0.630`, `min alpha 253`, identical to CPU float32.

**2. End to end through the full pipeline.** Confirmed on the complete `QwenImage21Pipeline` image-edit path (MPS, bfloat16, 768×1376, one reference image), not just the isolated VAE:

| | before fix | after fix | source image |
|---|---|---|---|
| RGB std | 31.0 / 22.2 | **56.7 / 57.1** | ~55 |
| mean saturation | 0.189 / 0.203 | **0.649 / 0.673** | 0.628 |
| output mode | RGBA, min alpha 173 | **RGB, fully opaque** | — |

Two seeds at 20 steps; a 40-step run behaves the same. Before the fix the edit was already *semantically* correct — the requested change was applied and the composition preserved — which is what makes the failure easy to misread as a prompt-adherence or guidance problem rather than a VAE one.

**3. The alpha anomaly resolves itself.** Once the temporal pad is correct, the decoded alpha returns to fully opaque, so downstream logic that infers transparency from the decoded alpha stops misfiring with no separate alpha handling.

The workaround also works applied at runtime as a monkeypatch on `QwenImage21AvgDown3D.forward`, gated on `device == "mps"`, so it can be validated without modifying the model code.

This is ultimately a PyTorch MPS bug and I intend to report it there too, but since the workaround is a one-liner and the current behaviour silently corrupts every image-conditioned generation on Apple Silicon, it seems worth fixing here as well.

## Environment

- Apple M5 Max, 128 GB unified memory, macOS (Darwin 25.6.0)
- PyTorch 2.14.0, MPS backend, `PYTORCH_ENABLE_MPS_FALLBACK=1`
- diffusers 0.41.0.dev0
- `Qwen/Qwen-Image-2.1`, VAE `AutoencoderKLQwenImage21` (`in_channels: 4`, `z_dim: 64`)
- Reproduced in `float32`, `float16` and `bfloat16`
